src/FractionalWrapper.sol

```diff
@@ -0,0 +1,112 @@
+    constructor(IERC20 asset_, string memory tokenName, string memory tokenSymbol) ERC20Mock(tokenName, tokenSymbol) {
+        asset = asset_;
+    }
+
+
+    /// @notice User to deposit underlying tokens for FractionalWrapper tokens
+    /// @dev Mints shares Vault shares to receiver by depositing exactly amount of underlying tokens
+    /// @param assets The amount of underlying tokens to deposit
+    /// @param receiver Address of receiver of Fractional Wrapper shares
+    function deposit(uint assets, address receiver) external returns(uint shares) {       
+        shares = convertToShares(assets);
+
+        //transfer DAI from user
+        bool success = asset.transferFrom(receiver, address(this), assets);
+        require(success, "Deposit failed!");   
+
+        //mint yvDAI to user
+        bool sent = _mint(receiver,convertToShares(assets));
+        require(sent, "Mint failed!"); 
+
+        emit Deposit(msg.sender, receiver, assets, shares);  
+    }
+
+
+    /// @notice User burns wrapped tokens, receiving udnerlying tokens; based on the exchange rate.
+    /// @dev Burns shares from owner and sends exactly assets of underlying tokens to receiver
+    /// @param assets The amount of underlying tokens to withdraw
+    /// @param receiver Address of receiver of underlying tokens
+    /// @param owner Address of owner of Fractional Wrapper shares
+    function withdraw(uint assets, address receiver, address owner) external returns(uint shares) {
+        shares = convertToAssets(assets);
+        
+        // MUST support a withdraw flow where the shares are burned from owner directly where owner is msg.sender,
+        // OR msg.sender has ERC-20 approval over the shares of owner
+        if(msg.sender != owner){
+            uint allowedShares = _allowance[owner][receiver] ;
+            require(allowedShares >= shares, "Allowance exceeded!");
+            _allowance[owner][receiver] = allowedShares - shares;
+        }
+        
+        //burn yvDAI 
+        burn(owner, shares);
+        
+        //transfer assets
+        bool success = asset.transfer(receiver, assets);
+        require(success, "Transfer failed!"); 
+
+        emit Withdraw(msg.sender, receiver, owner, assets, shares);
+    }
+
+
+    ///@dev set exchange rate 
+    function setExchangeRate(uint exRate_) public {
+        exRate = exRate_;
+    }
+
+    ///@notice calculate how much yvDAI user should get based on exchange rate
+    ///@dev exRate(27 dp precision) | both assets and shares are 18 dp precision
+    function convertToShares(uint assets) public view returns(uint shares){
+        return (assets)*(exRate / 10**27);
+    }
+
+    ///@notice calculate how much DAI user should get based on exchange rate
+    ///@dev exRate(27 dp precision) | both assets and shares are 18 dp precision
+    function convertToAssets(uint shares) public view returns(uint assets){
+        return (shares)/(exRate / 10**27);
```

Comment:
The arithmetic is not right, it's not the reverse of `convertToShares`. Have another go.